src/layers.jl

```diff
@@ -108,19 +125,32 @@ end
 
 @functor Patching
 
-"""
-    mlpblock(planes, expansion_factor = 4, dropout = 0., dense = Dense)
+# Positional embedding layer used by many vision transformer-like models
+struct PosEmbedding
+  embedding_vector
+end
 
-Feedforward block used in many vision transformer-like models.
+(p::PosEmbedding)(x) = x .+ p.embedding_vector[:, 1:size(x)[2], :]
 
-# Arguments
-  `planes`: Number of dimensions in the input and output.
-  `hidden_planes`: Number of dimensions in the intermediate layer.
-  `dropout`: Dropout rate.
-  `dense`: Type of dense layer to use in the feedforward block.
-  `activation`: Activation function to use.
-"""
-function mlpblock(planes, hidden_planes, dropout = 0., dense = Dense; activation = gelu)
-  Chain(dense(planes, hidden_planes, activation), Dropout(dropout),
-        dense(hidden_planes, planes, activation), Dropout(dropout))
+@functor PosEmbedding
+
+# Class tokens used by many vision transformer-like models
+struct CLSTokens
+  cls_token
 end
+
+function(m::CLSTokens)(x)
+  cls_tokens = repeat(m.cls_token, 1, 1, size(x)[3])
+  x = cat(cls_tokens, x; dims = 2)
+end
+
+@functor CLSTokens
+
+# Utility function to decide if mean pooling happens inside the model
+struct CLSPooling
+  mode
+end
+
+(m::CLSPooling)(x) = (m.mode == "cls") ? x[:, 1, :] : _seconddimmean(x)
+
+@functor CLSPooling
```

Comment:
This doesn't need to happen every forward pass right? Instead of introducing a new type, we can put `x -> x[:, 1, :]` or `x -> _seconddimmean(x)` in the chain during the `vit` function.

Reply:
I'm still relatively new to Julia so I'm just gonna express my amazement at how an anonymous function just integrates that neatly into the Chain API. That's just amazing🤯